src/snippet/mod.rs

```diff
@@ -0,0 +1,207 @@
+use tokenizer::{TokenStream, Tokenizer, Token};
+use std::collections::BTreeMap;
+use Term;
+use Document;
+use Index;
+use schema::FieldValue;
+use schema::Value;
+use tokenizer::BoxedTokenizer;
+
+#[derive(Debug)]
+pub struct HighlightSection {
+    start: usize,
+    stop: usize,
+}
+
+impl HighlightSection {
+    fn new(start: usize, stop: usize) -> HighlightSection {
+        HighlightSection {
+            start,
+            stop
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct FragmentCandidate {
+    score: f32,
+    start_offset: usize,
+    stop_offset: usize,
+    num_chars: usize,
+    highlighted: Vec<HighlightSection>,
+}
+
+impl FragmentCandidate {
+
+    fn new(start_offset: usize, end_offset: usize) -> FragmentCandidate {
+        FragmentCandidate{score: 0.0,
+                          start_offset: start_offset,
+                          stop_offset: end_offset,
+                          num_chars: 0,
+                          highlighted: vec![]}
+    }
+
+    /// Updates `score` and `highlighted` fields of the objects.
+    ///
+    ///
+    fn calculate_score(&mut self, token: &Token, terms: &BTreeMap<String, f32>) {
```

Comment:
This should be called try_add_token`.

The comment above should explain that if the token is one of the keyword, then the fragment length will be extended.